### PR TITLE
Firmware upgrade over USB Mass Storage

### PR DESCRIPTION
**Pull request by gsmcmullin**
Implementation of https://github.com/swift-nav/firmware_team_planning/issues/247

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## v0.2-5-g87460c8
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v0.2-5-g87460c8/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v0.2-5-g87460c8/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v0.2-5-g87460c8

File: package/sbp_log/src/main.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+#include <libsbp/sbp.h>
+#include <libsbp/logging.h>
+#include <czmq.h>
+#include <getopt.h>
+#include <unistd.h>
+
+#define SBP_FRAMING_MAX_PAYLOAD_SIZE 255
+static u16 sender_id = SBP_SENDER_ID;
```
Comment (denniszollo) [inline package/sbp_log/src/main.c]: sender_id should come from lowest 16 bits of the uuid from the factory data (not hard coded)

**Comment by denniszollo** [comment]
This did not work for me; the device seemed to boot up normally and didn't do a firmware upgrade after plugging in the ubs drive.

After linux came up, I was able to see the file mounted.
`>>> print fi.readdir('/media/sda1/')
['.', '..', '._.Trashes', '.Trashes', '.Spotlight-V100', '.fseventsd', 'PiksiMulti-v0.2-2.bin', '._PiksiMulti-v0.2-2.bin', 'System Volume Information']`

Noticeably I didn't see a log created for the upgrade process in /var/log

print fi.readdir('/var/log')
['.', '..', 'messages', 'messages.0', 'monit.err', 'monit.log', 'zmq_adapter_uart_ttyPS1.err', 'zmq_adapter_uart_ttyPS1.log', 'zmq_adapter_uart_ttyPS0.err', 'zmq_adapter_uart_ttyPS0.log', 'zmq_adapter_tcp_listen.err', 'zmq_adapter_tcp_listen.log', 'zmq_adapter_rpmsg_piksi100.err', 'zmq_adapter_rpmsg_piksi100.log', 'zmq_adapter_uart_ttyGS0.err', 'zmq_adapter_uart_ttyGS0.log', 'piksi_system_daemon.err', 'piksi_system_daemon.log', 'sbp_settings_daemon.err', 'sbp_settings_daemon.log', 'zmq_router.err', 'zmq_router.log', 'sshd']

**Comment by denniszollo** [comment]
Update:

When I started from a new DEV configuration, loading this linux image off of sd card, the upgrade proceeded and worked from a usb flash drive.  I got a few flash read/write errors that are a little troubling shown below.
```
U¥^@B^@0ethernet^@ip_config_mode^@Static^@enum:Static,DHCP^@â<8d>IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15740 U¥^@B^@0ethernet^@ip_config_mode^@Static^@enum:Static,DHCP^@â<8d>U^A^DB^@?^FNew firmware image set detected: /media/sda1/PiksiMulti-*.bin
15741 <95>[IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15742 U¥^@B^@0ethernet^@ip_config_mode^@Static^@enum:Static,DHCP^@â<8d>U^A^DB^@^W^FPerforming upgrade...
15743 õ<9e>IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15744 U¥^@B^@0ethernet^@ip_config_mode^@Static^@enum:Static,DHCP^@â<8d>IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15745 U¥^@B^@0ethernet^@ip_config_mode^@Static^@enum:Static,DHCP^@â<8d>IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15746 U¥^@B^@"ethernet^@ip_address^@192.168.0.222^@{øIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15747 U¥^@B^@"ethernet^@ip_address^@192.168.0.222^@{øIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15748 U¥^@B^@"ethernet^@ip_address^@192.168.0.222^@{øIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15749 U¥^@B^@"ethernet^@ip_address^@192.168.0.222^@{øIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15750 U¥^@B^@"ethernet^@ip_address^@192.168.0.222^@{øIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15751 U¥^@B^@^_ethernet^@netmask^@255.255.255.0^@ò
15752 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15753 U¥^@B^@^_ethernet^@netmask^@255.255.255.0^@ò
15754 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15755 U¥^@B^@^_ethernet^@netmask^@255.255.255.0^@ò
15756 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15756 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15757 U¥^@B^@^_ethernet^@netmask^@255.255.255.0^@ò
15758 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15759 U¥^@B^@^_ethernet^@netmask^@255.255.255.0^@ò
15760 IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15761 U¥^@B^@^]ethernet^@gateway^@192.168.0.1^@^HáIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15762 U¥^@B^@^]ethernet^@gateway^@192.168.0.1^@^HáIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15763 U¥^@B^@^]ethernet^@gateway^@192.168.0.1^@^HáIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15764 U¥^@B^@^]ethernet^@gateway^@192.168.0.1^@^HáIPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready^M
15765 U¥^@B^@^]ethernet^@gateway^@192.168.0.1^@^Hámacb e000b000.ethernet eth0: link up (10/Full)^M
15766 IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready^M
15767 macb e000b000.ethernet eth0: link down^M
15768 macb e000b000.ethernet eth0: link up (10/Full)^M
15769 macb e000b000.ethernet eth0: link down^M
15770 mmcblk0: error -84 transferring data, sector 24576, nr 2, cmd response 0x900, card status 0xc00^M
15771 mmc0: tried to reset card^M
15772 mmcblk0: error -84 transferring data, sector 24576, nr 2, cmd response 0x900, card status 0xc00^M
15773 blk_update_request: I/O error, dev mmcblk0, sector 24576^M
15774 Buffer I/O error on dev mmcblk0p1, logical block 16384, lost async page write^M
15775 blk_update_request: I/O error, dev mmcblk0, sector 24577^M
15776 Buffer I/O error on dev mmcblk0p1, logical block 16385, lost async page write^M
15777 macb e000b000.ethernet eth0: link up (10/Full)^M
15778 macb e000b000.ethernet eth0: link down^M
15779 U^A^DB^@^X^Finstalling to region A
15780 ýôU^A^DB^@)^Ferasing mtd2 0x00020000 - 0x00030000...
15781 üÈU^A^DB^@^D^Fok
15782 ^^6U^A^DB^@)^Ferasing mtd3 0x00080000 - 0x000c0000...
15783 .±U^A^DB^@^D^Fok
15784 ^^6U^A^DB^@)^Ferasing mtd4 0x00000000 - 0x03d00000...
15785 ]U^A^DB^@^D^Fok
15786 ^^6U^A^DB^@)^Fwriting mtd4 0x00000000 - 0x0006bc3c...
15787 *ZU^A^DB^@^D^Fok
15788 ^^6U^A^DB^@)^Fwriting mtd4 0x0006bc40 - 0x004477eb...
15789 ¤\U^A^DB^@^D^Fok
15790 ^^6U^A^DB^@)^Fwriting mtd3 0x00080000 - 0x0009446b...
15791 ^A?U^A^DB^@^D^Fok
15792 ^^6U^A^DB^@)^Fwriting mtd2 0x00020000 - 0x00020900...
15793 xwU^A^DB^@^D^Fok
15794 ^^6U^A^DB^@ ^Fupgrade completed successfully
```

**Comment by jacobmcnamee** [comment]
Thoughts:
- If the USB drive isn't reliably mounting before this script is called, should we poll for that?
- Would it be better to run the whole script in the background and potentially delay ~10s? Or just poll continuously every so often?

File: package/sbp_log/src/main.c (hunk 1 above)
Comment (jacobmcnamee) [inline package/sbp_log/src/main.c]: FYI `uuid` may be read from `/factory/uuid`

File: package/sbp_log/src/main.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  /* Delay for long enough for socket thread to sort itself out */
+  usleep(100000);
+
+  while (fgets(msg->text,
+               SBP_FRAMING_MAX_PAYLOAD_SIZE - offsetof(msg_log_t, text),
+               stdin)) {
+    sbp_send_msg(&sbp, SBP_MSG_LOG, sizeof(*msg) + strlen(msg->text), (u8*)msg);
+  }
+
+  zsock_destroy(&zpub);
```
Comment (jacobmcnamee) [inline package/sbp_log/src/main.c]: Do we need any kind of flush or delay to make sure everything has been sent before closing the socket?

File: board/piksiv3/rootfs-overlay/etc/init.d/S84upgrade_tool [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  fi
+
+  if [ ! -x $FIRMWARE ]; then
+    exit
+  fi
+
+  echo "New firmware image set detected: $FIRMWARE" | sbp_log $LOGLEVEL
+  echo "Performing upgrade..." | sbp_log $LOGLEVEL
+  upgrade_tool --debug $FIRMWARE | sbp_log $LOGLEVEL
+  rm $FIRMWARE
```
Comment (jacobmcnamee) [inline board/piksiv3/rootfs-overlay/etc/init.d/S84upgrade_tool]: Maybe it would be better to rename instead of deleting?

**Comment by denniszollo** [comment]
@jacobmcnamee 

I don't think it's better to run whole script in the bkg and delay 10 seconds because I think it's better to explicitly NOT Boot up the firmware and navigation if an upgrade is to commence.  If we could kill piksi_firmware when we decide to start the upgrade, I think this could be an option.  Running the firmware during the upgrade adds complexity, makes it hard to tell user how things are progressing, and sometimes prevents flash read writes leading to errors.

I wouldn't want to constantly poll for a flash drive as this might make future logging functionality harder and could lead to an inadvertent upgrade. I also think this is a worse user experience.

/cc @jdormoy @gsmcmullin @switanis

**Comment by anonymous-travis-swiftnav** [comment]
## v0.2-20-g683a307
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v0.2-20-g683a307/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v0.2-20-g683a307/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v0.2-20-g683a307

**Pull request closed without merging**